src/plugins/legacy/polygonRoi/toolboxes/polygonRoiToolBox.cpp

```diff
@@ -432,32 +438,24 @@ void polygonRoiToolBox::updateTableWidgetView(unsigned int row, unsigned int col
             handleDisplayError(medAbstractProcessLegacy::FAILURE);
             return;
         }
-        for (int i=0; i<mainView->layersCount(); i++)
+        for (unsigned int i=0; i<mainView->layersCount(); i++)
         {
             container->addData(mainView->layerData(i));
         }
 
+        // Closing behavior of the non-main view
         medAbstractImageView* view = static_cast<medAbstractImageView *> (container->view());
-        connect(view, &medAbstractImageView::closed, [this, mainContainer](){
-            if (viewEventFilter)
+        connect(view, &medAbstractImageView::closed, [this, mainContainer]()
+        {
+            if (viewEventFilter && mainContainer)
             {
                 viewEventFilter->clearAlternativeView();
-                medTabbedViewContainers *tabs = this->getWorkspace()->tabbedViewContainers();
-                if (tabs)
-                {
-                    QList<medViewContainer*> containersInTab = tabs->containersInTab(tabs->currentIndex());
-                    if (containersInTab.size()==1 && containersInTab.at(0)->uuid()==mainContainerUUID)
-                    {
-                        if (activateTBButton->isEnabled())
-                        {
-                            repulsorTool->setEnabled(true);
-                            tableViewChooser->setEnabled(true);
-                        }
-                    }
-                }
```

Comment:
I do not understand why you remove this lines.
In some scenario, we need to enable the repulsor and the tableView. You can not just remove it, no ?
Or maybe I missed something ?

Reply:
Maybe i was not doing the right tests, but i did not see any case where you need to enable the repulsor and orientation buttons at the closing of the alternate view. When you split the views with the orientation buttons, and you remove the alternate view, the buttons on the original view are good. Is there an other case where you would need it?

Reply:
I can not remember rigth now.
But I do not write this code without reason. 
It was to answer to a particular scenario useful for users.
As I said before, I have to develop new features in this toolbow soon. 
I have planned to do a refactoring at this time.

Reply:
Every crashs related to open/close view and/or containers must not be fixed in the toolbox.
It is a deeper issue in the way of managing the views and containers in core I think.

Reply:
I think i know the origin of this code: in order to avoid crashs, when the user choose to use the split button, the main view is blocked, meaning you can't remove it. If you remove the second view, the main view switch to the closable mode. However, before this PR you could remove the main view even with split views, meaning that at the removal of the first view you needed to use the alternate view as a new main view. This code is not needed anymore

Reply:
ok but i made some actions on repulsor for instance
Are you sure the state of the repulsor is ok now ?

Reply:
And It is important to keep the first view as a main view because the segmentation is made in this orientation.
Not sure, you can continue to segment the images without the first view.

Reply:
I have read this post about C++ lambda expressions and Qt connect mechanism 
https://medium.com/genymobile/how-c-lambda-expressions-can-improve-your-qt-code-8cd524f4ed9f

This comments seems interesting:
" Qt connection system is smart enough to delete connections if either the sender or the receiver is deleted, so in our first version of setMonitor(), if monitor is deleted the connection is deleted as well. But now we use a lambda for the receiver: Qt has no way to now that this lambda makes use of monitor. Even if monitor is deleted, the lambda will still be called and the application will crash when it tries to dereference monitor.

To avoid that, you can pass a “context” argument to the connect() call, [...]
we pass monitor as a context to connect(). It won't affect the execution of our lambda, but when monitor is deleted, Qt will notice and will disconnect Worker::progress() from our lambda."

Reply:
Also, it is not useful to test `mainContainer` because it is captured by value in the lambda expression (see https://thispointer.com/c11-lambda-how-to-capture-local-variables-inside-lambda/), and you cannot capture it by reference since it is a local pointer.

Reply:
Great, if the main view should not change, it amplifies the need to lock the main view if the view is split.

I need to add that the lambda connection is done on `view` which is from `container->view()`, which is the split view, not the main one. So, this connection is only used at the closing on the second view. So, it means that at the closing of the second view, if there is only the main view, we want to activate the repulsor and split buttons. However, it is already done without this code.

I'm thinking about something else also: in order to avoid any problems, it would be interesting in the future not to allow the view to be split in the Tools button. Only the split buttons of the toolbox should be accepted i think.

Thanks for the documentation, it's very interesting. If i understand well, if we would like to keep the code removed here, or for a future code, we could pass a context variable. Here, i'm not sure which one, for instance `this->getWorkspace()` crash in it at the closing of the application. Would it be possible to pass `this` or a function?

Reply:
(I edited the previous comment because you can technically get a reference (by dereferencing), but the test would still be pointless since the reference will not change. What I meant was you shouldn't be testing `mainContainer`like this but instead call `containersInTabSelected.at(0)`directly inside the lambda if you want to check that the container still exists when the lambda is called)

Reply:
Oki, i'll remove the test on `mainContainer`. However, we could not test directly (without the use of context param) `containersInTabSelected` because `this->getWorkspace()->tabbedViewContainers();` crashes before that

Reply:
(our posts crossed) The context should be `this` and `mainContainer` should not be captured but accessed using `containersInTabSelected.at(0)`. That way it is not important who gets deleted first: if it is the toolbox (i.e. `this`) then the lambda will be disconnected, and if there is no container then you will get a null value (must test it of course).

Reply:
Thanks for your help, indeed it does not crash at the end of the application anymore, even with the call to `getWorkspace()`! What do you want to do with the lines:

```c++
                    if (containersInTab.size()==1 && containersInTab.at(0)->uuid()==mainContainerUUID)
                    {
                        if (activateTBButton->isEnabled())
                        {
                            repulsorTool->setEnabled(true);
                            tableViewChooser->setEnabled(true);
                        }
                    }
```
They seems not to be used anymore. We can remove them, or put them back and i can add a comment saying that they seems obsolete, or i can put them back. As you want

Reply:
I prefer the second choice if possible.
I will work on it in few weeks

Reply:
Done :)